react/features/conference/middleware.js

```diff
@@ -48,3 +51,25 @@ MiddlewareRegistry.register(store => next => action => {
 
     return result;
 });
+
+/**
+ * Set up state change listener to perform maintenance tasks when the conference
+ * is left or failed, close all dialogs and unpin any pinned participants.
+ */
+StateListenerRegistry.register(
+    state => getCurrentConference(state),
+    (conference, { dispatch, getState }, prevConference) => {
+        const { authRequired, passwordRequired }
+            = getState()['features/base/conference'];
+
+        if (conference !== prevConference) {
+            dispatch(pinParticipant(null));
+
+            if (typeof authRequired === 'undefined'
+                    && typeof passwordRequired === 'undefined') {
+                // Conference changed, left or failed... and there is no
+                // pending authentication, close any dialog we might have open.
+                dispatch(hideDialog());
```

Comment:
There was a slight discussion recently about what should feature middlewares handle and what not. Although we didn't have a clear consensus, this seems like it belongs to the dialog middleware to me. What do you think?

Mentioning @virtuacoplenny as we had the conversation with him. What do you think Lenny?

Reply:
I don't think so. The dialog itself doesn't know when dialogs should be shown or not. The conference decides when a dialog should be shown, and incidentally, when it should be hidden.

Reply:
Yeah, this is exactly why there can be different views on the topic.

In my world view features like```conference``` do their own things (for example they start, fail, finish...etc) and let the world (like ```dialogs```) react to what they do with their own internal state. And not the other way round, where the conference finishes and the conference should immediately know which features' state should be cleared up and which not.

Or do you think features should know about other features' behaviour and clear their state up?

Reply:
I don't think it's such a clean cut. We have features which are pretty self-contained: full-screen, for example. But features interact with each other and delegate responsibility: the `dialogs` feature doesn't know **when** they need to be shown or hidden, but it offers dialog capabilities to other features. `conference` OTOH, doesn't know about dialog state, but uses the tools available to itself in order to open / close dialogs when the time is right.

Reply:
Ok, I'm not against this change, just adding my 2 cents. So one last question, then I approve it (with my other question above in mind please).

If the dialogs have to be forcedly closed also when there is some other action happening (not sure about an idea now, but I think you can imagine), then this general ```dispatch(hideDialog())``` will be injected into the code on multiple locations? Isn't it better to say that we list all action types when the dialogs need to be closed in the dialogs middleware and that handles it all on one location?

Reply:
No. It's the same as the idea of recoverable errors: there isn't a central place which decides what errors are recoverable and what aren't, features which can recover an error will set the recoverable flag and move on.

In general, we have a few patterns coexisting here, since there are many different use cases, and while I see the appeal to try to unify all the things, at the end of the day if would complicate our lives because of too tight of a coupling.